reorganize Pattern library

guberathome opened the pull request:
* feat!: Pattern.match now returns start and end indizes of the first match
* feat!: matches from Pattern.find-all can no longer overlap. This fixes #1255.
* feat!: Pattern.global-match-str now returns all matches as String 
* feat!: renamed (Pattern.global-match) to (Pattern.match-all-groups) 
* feat: implemented (Pattern.match-from)
* feat: moved Pattern.match-str and Pattern.find from C to Carp code 
* fix: moved some matching code from C to Carp and removed dead C functions

hellerve commented on core/Pattern.carp:
```diff
@@ -1,6 +1,15 @@
 
 (defmodule Pattern
+  (register-type MatchResult "PatternMatchResult" [start Int, end Int])
+  (defn       match2str [match-res] 
+	(fmt "#<MatchResult start=%d end=%d>" 
+	  (MatchResult.start match-res) 
+	  (MatchResult.end   match-res) ))
+  (implements str Pattern.match2str)
```
The convention here would be to do

```clojure
(defmodule MatchResult
  (defn str [match-res]
    ; your code
  )
  (implements str PatternMatchResult.str)
)
```

in a nested module, unless `str` clashes with anything.

guberathome replied:
Thank you for the pointer, updated the sources.

Printing a MatchResult in the REPL works, but printing (Array.MatchResult) does not.
What am I missing?
```
> (info Pattern.MatchResult.refstr)
Pattern.MatchResult.refstr : (Fn [(Ref Pattern.MatchResult a)] String)
  Defined at line 8, column 6 in '/home/ptokremin/contrib_ghub/Carp/core/Pattern.carp'
  Implements: (prn str)
> (info Pattern.MatchResult.str)
Pattern.MatchResult.str : (Fn [Pattern.MatchResult] String)
  Defined at line 16, column 6 in '/home/ptokremin/contrib_ghub/Carp/core/Pattern.carp'
  Documentation: converts a `Pattern.MatchResult` to a string.
  Implements: (prn str)
> (Pattern.match    #"<(.-)>{(.-)}" "Whatever we <do>{cmd1} is going to cause some <trouble>{msg2} no matter...")
(MatchResult start=12 end=22)
> (info Pattern.find-all-matches )
Pattern.find-all-matches : (Fn [(Ref Pattern a), (Ref String b)] (Array Pattern.MatchResult))
  Defined at line 47, column 4 in '/home/ptokremin/contrib_ghub/Carp/core/Pattern.carp'
> (Pattern.find-all-matches     #"<(.-)>{(.-)}" "Whatever we <do>{cmd1} is going to cause some <trouble>{msg2} no matter...")
[Warning] No 'prn' function found with type (Fn [Pattern.MatchResult] String).
carp: Can't find any 'prn' function for member 'Inside array.' of type (Fn [Pattern.MatchResult] String)
CallStack (from HasCallStack):
  error, called at src/ArrayTemplates.hs:634:33 in CarpHask-0.5.1.0-EudeCbDnX8j3pnSP31uN8I:ArrayTemplates
```

hellerve replied:
No, the `prn` function required here is of type `(Fn [Pattern.MatchResult] String)`. You defined one of type `(Fn [&Pattern.MatchResult] String)`.

guberathome replied:
Sorry, I omited that bit for brevity sake. Here's the complete module:
```
(defmodule Pattern

  (register-type MatchResult "PatternMatchResult" [start Int, end Int])
  (defmodule MatchResult
    (defn refmatch2str [ref-matchres] 
      (fmt "(MatchResult start=%d end=%d)" 
        (MatchResult.start ref-matchres) 
        (MatchResult.end   ref-matchres) ))
    (implements str Pattern.MatchResult.refmatch2str)
    (implements prn Pattern.MatchResult.refmatch2str)
    ;
    (defn match2str [matchres] 
      (Pattern.MatchResult.refmatch2str &matchres) )
    (implements str Pattern.MatchResult.match2str)
    (implements prn Pattern.MatchResult.match2str)
  )
  ; ... code
)
```

Functions look ok from the transcript:
```
> (info Pattern.MatchResult.match2str)
Pattern.MatchResult.match2str : (Fn [Pattern.MatchResult] String)
  Defined at line 14, column 6 in '/home/ptokremin/contrib_ghub/Carp/core/Pattern.carp'
  Implements: (prn str)
> (info Pattern.MatchResult.refmatch2str)
Pattern.MatchResult.refmatch2str : (Fn [(Ref Pattern.MatchResult a)] String)
  Defined at line 7, column 6 in '/home/ptokremin/contrib_ghub/Carp/core/Pattern.carp'
  Implements: (prn str)
> (Pattern.match    #"<(.-)>{(.-)}" "Whatever we <do>{cmd1} is going to cause some <trouble>{msg2} no matter...")
(MatchResult start=12 end=22)
> (Pattern.find-all-matches     #"<(.-)>{(.-)}" "Whatever we <do>{cmd1} is going to cause some <trouble>{msg2} no matter...")
[Warning] No 'prn' function found with type (Fn [Pattern.MatchResult] String).
carp: Can't find any 'prn' function for member 'Inside array.' of type (Fn [Pattern.MatchResult] String)
CallStack (from HasCallStack):
  error, called at src/ArrayTemplates.hs:634:33 in CarpHask-0.5.1.0-EudeCbDnX8j3pnSP31uN8I:ArrayTemplates
```

Could this be related to nested modules?

hellerve replied:
No, this seems fine now. This might possibly be due to the old array code still looking for a function named `prn` rather than a function marked as implementing `prn`. Maybe @scolsen (who built the `implements` mechanism) has some insight on this?

guberathome replied:
Ok, the problem is elsewhere. I'll open a separate issue.

guberathome commented:
What is the purpose of having ```(Pattern.global-match-str)``` _and_ ```(Pattern.global-match)```? 
Is it to match _all_ Groups??
Could we remove or rename ```(Pattern.global-match)```?

```
> (Pattern.global-match-str   #"'.-'" "Whatever we 'do' is going to cause some 'trouble' no matter...")
[@"'do'" @"'trouble'"]
> (Pattern.global-match   #"'.-'" "Whatever we 'do' is going to cause some 'trouble' no matter...")
[[@"'do'"] [@"'trouble'"]]
```

hellerve commented:
The rename seems good to me!

guberathome commented:
done, Pattern.split now works as intended:
```
> (Pattern.global-match-str   #"<(.-)>{(.-)}" "Whatever we <do>{cmd1} is going to cause some <trouble>{msg2} no matter...")
[@"<do>{cmd1}" @"<trouble>{msg2}"]
> (Pattern.find-all   #"<(.-)>{(.-)}" "Whatever we <do>{cmd1} is going to cause some <trouble>{msg2} no matter...")
[12 46]
> (Pattern.split   #"<(.-)>{(.-)}" "Whatever we <do>{cmd1} is going to cause some <trouble>{msg2} no matter...")
[@"Whatever we " @" is going to cause some " @" no matter..."]
> (Pattern.match-all-groups     #"<(.-)>{(.-)}" "Whatever we <do>{cmd1} is going to cause some <trouble>{msg2} no matter...")
[[@"do" @"cmd1"] [@"trouble" @"msg2"]]
```

eriksvedang commented on core/Pattern.carp:
```diff
@@ -1,28 +1,80 @@
 
 Returns `[]` if it doesn’t find a matching pattern.")
-  (register find-all (Fn [&Pattern &String] (Array Int)))
+  ;~ (register find-all (Fn [&Pattern &String] (Array Int)))
```
Remove this line?

guberathome replied:
done

eriksvedang commented on core/Pattern.carp:
```diff
@@ -1,28 +1,80 @@
+        (MatchResult.end   ref-matchres) ))
+    (implements str Pattern.MatchResult.refmatch2str)
+    (implements prn Pattern.MatchResult.refmatch2str)
+    ;
```
Remove ; ?

guberathome replied:
the trailing ";"? 
It's gone

eriksvedang commented on core/Pattern.carp:
```diff
@@ -1,28 +1,80 @@
+
+  (register-type MatchResult "PatternMatchResult" [start Int, end Int])
+  (defmodule MatchResult
+    (defn refmatch2str [ref-matchres] 
```
I'd prefer "refmatch-to-str". This "2" naming scheme is not something we use anywhere else in core.

TimDeve replied:
More idiomatic names would be `match2str` => `str`, `refmatch2str` => `ref-str`. You don't really want to repeat the name of the type while in the associated module.

guberathome replied:
done

eriksvedang reviewed (commented):
Really nice improvements! Had some nitpicks, but the important parts look great.
